python/lsst/meas/astrom/pessimistic_pattern_matcher_b_3D.py

```diff
@@ -0,0 +1,1111 @@
+                        src_sin_tol**2 < src_sin_tol**2):
+                    continue
+
+            # The cosine tests the manitude of the angle but not
+            # its direction. To do that we need to know the sine as well.
+            # This cross product calculation does that.
+            cross_ref = ref_sign*(
+                np.cross(ref_delta_array[ref_dist_idx], ref_delta) /
+                (ref_dist_array[ref_dist_idx]*ref_dist))
+            dot_cross_ref = np.dot(cross_ref, ref_ctr)
+
+            # Check the value of the cos again to make sure that it is not
+            # near zero.
+            sin_comp = -src_sin_tol < cos_theta_src < src_sin_tol
+            # Test the difference of the sings of the reference angle against
+            # the source angle. Assumes that the delta between the two is
+            # small.
+            if sin_comp:
+                if not (-src_sin_tol <
+                        (dot_cross_src - dot_cross_ref) /
+                        src_sin_tol < src_sin_tol):
+                    continue
+            else:
+                if not (-src_sin_tol <
+                        (dot_cross_src - dot_cross_ref) /
+                        cos_theta_ref < src_sin_tol):
+                    continue
+
+            # Return the correct id depending on the direction of the vector.
+            if ref_sign == 1:
+                return ref_id_array[ref_dist_idx, 1]
+            else:
+                return ref_id_array[ref_dist_idx, 0]
+        return None
+
+    def _create_shift_rot_matrix(self, cos_rot_sq, shift_matrix, src_delta,
+                                 ref_ctr, ref_delta):
+        """ Create the final part of our spherical rotation matrix.
+
+        Parameters
+        ----------
+        cos_rot_sq : float
+            cosine of the rotation needed to align our source and reference
+            candidate patterns.
+        shift_matrix : float array
+            3x3 rotation matrix for shifting the source pattern center on top
+            of the candidate reference pattern center.
+        src_delta : float array
+            3 vector delta of representing the first spoke of the source
+            pattern
+        ref_ctr : float array
+            3 vector on the unitsphere representing the center of our
+            reference pattern.
+        ref_delta : float array
+            3 vector delta made by the first pair of the reference pattern.
+
+        Returns
+        -------
+        lsst.pipe.base.Struct
+            Struct object containing the following data
+            sin_rot : float sine of the amount of rotation between the
+                source and reference pattern. We use sine here as it is
+                signed and tells us the chirality of the rotation.
+            shift_rot_matrix : float array representing the 3x3 rotation
+                matrix that takes the source patern and shifts and rotates
+                it to align with the reference pattern.
```

Comment:
Very nice descriptions, but I think each description needs to start on the line *after* the name and type